### PR TITLE
Sma 46 multichain validation module

### PR DESCRIPTION
**Pull request by livingrockrises**
# Description

Multichain validation module implementation for multi chain userops signing.
Actual test script would be part of test cases soon - and also in backend scripts modulat-SA-v2 playground.

can change the base later if other PRs are merged first.

**EDIT: This is done** 

I am trying to send one tx on bnb + polygon testnets soon using this module. 
later we can send the ops which actually makes uses of enabled session.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

I have tested from backend-node scripts in sdk-examples repo [here](https://github.com/bcnmy/sdk-examples/tree/modular-sa-v2-playground/backend-node)

```
let provider = new ethers.providers.JsonRpcProvider(config.rpcUrl);
  let signer = new ethers.Wallet(config.privateKey, provider);
  const eoa = await signer.getAddress();
  console.log(chalk.blue(`EOA address: ${eoa}`));

  // create bundler and paymaster instances
  const bundler1 = new Bundler({
    bundlerUrl: config.bundlerUrl,
    chainId: config.chainId,
    entryPointAddress: DEFAULT_ENTRYPOINT_ADDRESS,
  });

  const paymaster = new BiconomyPaymaster({
    paymasterUrl: config.biconomyPaymasterUrl
  });

  const multiChainModule = new MultiChainValidationModule({
    signer: signer,
    chainId: config.chainId, // to be removed
    moduleAddress: '0x2E817fe3749B81dA801fc08B247E081ec20eB080'
  })
  
  const biconomySmartAccountConfig1 = {
    signer: signer,
    chainId: config.chainId,
    rpcUrl: config.rpcUrl,
    paymaster: paymaster, 
    bundler: bundler1, 
    entryPointAddress: DEFAULT_ENTRYPOINT_ADDRESS,
    defaultValidationModule: multiChainModule,
    activeValidationModule: multiChainModule
  };

  // create biconomy smart account instance
  const biconomyAccount1 = new BiconomySmartAccountV2(biconomySmartAccountConfig1);

  // passing accountIndex is optional, by default it will be 0. You may use different indexes for generating multiple counterfactual smart accounts for the same user
  const biconomySmartAccount1 = await biconomyAccount1.init();


// For another chain instance

const bundler2 = new Bundler({
    bundlerUrl: "https://bundler.biconomy.io/api/v2/97/A5CBjLqSc.0dcbc53e-anPe-44c7-b22d-21071345f76a",
    chainId: 97,
    entryPointAddress: DEFAULT_ENTRYPOINT_ADDRESS,
  });

  const biconomySmartAccountConfig2 = {
    signer: signer,
    chainId: 97,
    rpcUrl: 'https://data-seed-prebsc-1-s2.binance.org:8545',
    // paymaster: paymaster, 
    bundler: bundler2, 
    entryPointAddress: DEFAULT_ENTRYPOINT_ADDRESS,
    defaultValidationModule: multiChainModule,
    activeValidationModule: multiChainModule
  };


  // create biconomy smart account instance
  const biconomyAccount2 = new BiconomySmartAccountV2(biconomySmartAccountConfig2);

  // passing accountIndex is optional, by default it will be 0. You may use different indexes for generating multiple counterfactual smart accounts for the same user
  const biconomySmartAccount2 = await biconomyAccount2.init();
  
  // Building user ops
  
  // any transaction object (say mintNFT)
  // or maybe enableModule (session key manager address)
  // or maybe enableModule and createSession batch 
  // createSession() tx data
  
  let partialUserOp1 = await biconomySmartAccount1.buildUserOp([transaction1]);
  
  let partialUserOp2 = await biconomySmartAccount2.buildUserOp([transaction2]);
  
  const returnedOps = await multiChainModule.signUserOps([{userOp: partialUserOp1, chainId: 80001}, {userOp: partialUserOp2, chainId: 97}]);
  
  // Sending individual ops with common multichain sig
  
  // Mumbai
  
  const userOpResponse1 = await biconomySmartAccount1.sendSignedUserOp(returnedOps[0] as any);
  console.log(chalk.green(`userOp Hash: ${userOpResponse1.userOpHash}`));
  const transactionDetails1 = await userOpResponse1.wait();
  
  
  // BNB testnet
  
  const userOpResponse2 = await biconomySmartAccount2.sendSignedUserOp(returnedOps[1] as any);
  console.log(chalk.green(`userOp Hash: ${userOpResponse2.userOpHash}`));
  const transactionDetails2 = await userOpResponse2.wait();
  
```

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/modules/src/MultichainValidationModule.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        throw new Error(`Invalid version ${moduleConfig.version}`)
+      }
+      this.moduleAddress = moduleAddr
+      this.version = moduleConfig.version as ModuleVersion
+    }
+  }
+
+  // TODO: Actually depends on the module address so maybe we can make it dynamic
+  getDummySignature(): string {
+    return '0x00000000000000000000000000000000000000000000000000000000000000400000000000000000000000002E817fe3749B81dA801fc08B247E081ec20eB080000000000000000000000000000000000000000000000000000000000000004181d4b4981670cb18f99f0b4a66446df1bf5b204d24cfcb659bf38ba27a4359b5711649ec2423c5e1247245eba2964679b6a1dbb85c992ae40b9b00c6935b02ff1b00000000000000000000000000000000000000000000000000000000000000'
```
Comment (AmanRaj1608) [inline packages/modules/src/MultichainValidationModule.ts:34]: Why is this needed, what needs to be updated?

File: packages/modules/src/MultichainValidationModule.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import MerkleTree from 'merkletreejs'
+import { ECDSAOwnershipValidationModule } from './ECDSAOwnershipValidationModule'
+import { MULTICHAIN_VALIDATION_MODULE_ADDRESSES_BY_VERSION } from './utils/Constants'
+import { keccak256, arrayify, defaultAbiCoder, hexConcat, hexZeroPad } from 'ethers/lib/utils'
+import {
+  ECDSAOwnershipValidationModuleConfig,
+  ModuleVersion,
+  MultiChainUserOpDto
+} from './utils/Types'
+export class MultiChainValidationModule extends ECDSAOwnershipValidationModule {
```
Comment (livingrockrises) [inline packages/modules/src/MultichainValidationModule.ts:13]: @AmanRaj1608 @tomarsachin2271 since it extends the ECDSAModule chainId becomes mandatory but in real sense this module itself doesn't need to maintain chainId as it will receive it as part of dto (array of objects - userop + chainId) in singUserOps method. 
what should I do?

File: packages/modules/src/MultichainValidationModule.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  // Review: this.chainId should not be in spec of this class
```
Comment (livingrockrises) [inline packages/modules/src/MultichainValidationModule.ts:14]: the actual comment is here

File: packages/modules/src/MultichainValidationModule.ts (hunk 1 above)
Comment (livingrockrises) [reply packages/modules/src/MultichainValidationModule.ts]: done

**Comment by livingrockrises** [comment]
also updated sample client side code.

**Review by AmanRaj1608 — COMMENTED** [review]
Can refactor the commented codes and merge

**Review by AmanRaj1608 — APPROVED** [review]

File: packages/modules/src/MultichainValidationModule.ts (hunk 2 above)
Comment (livingrockrises) [reply packages/modules/src/MultichainValidationModule.ts:13]: ok let me do that.

File: packages/modules/src/MultichainValidationModule.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    throw new Error('Method not implemented.')
+  }
+
+  async signUserOps(multiChainUserOps: MultiChainUserOpDto[]): Promise<UserOperation[]> {
+    const leaves: string[] = []
+
+    // TODO
+    const validUntil = 0 // unlimited
+    const validAfter = 0
+
```
[the remark this answers is not in the document]
Comment (livingrockrises) [reply packages/modules/src/MultichainValidationModule.ts]: done
